src/api/tryGetFileTemporaryUrl.js

```diff
@@ -1,19 +1,49 @@
 /* @flow strict-local */
 import type { Auth } from './apiTypes';
+import { isUrlOnRealm } from '../utils/url';
 import getFileTemporaryUrl from './messages/getFileTemporaryUrl';
 
 /**
- * Get the complete authless URL to a realm uploaded file, returns null if that
- * fails.
+ * Like getFileTemporaryUrl, but on error returns null instead of throwing.
  *
- * See aslo jsdoc on getFileTemporaryUrl.
- *
- * @param href URL path to the resource.
- * @param auth Current user's auth information.
+ * Validates `url` to give getFileTemporaryUrl the kind of input it expects
+ * and returns null if validation fails.
  */
-export default async (href: string, auth: Auth): Promise<string | null> => {
+export default async (url: URL, auth: Auth): Promise<URL | null> => {
+  if (!isUrlOnRealm(url, auth.realm)) {
+    return null;
+  }
+
+  // A "path-absolute-URL string".
+  //
+  // When the unparsed `href` is a relative URL (i.e., one meant for the
+  // realm), we've seen it with and without a leading slash:
+  //   https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/Interpreting.20links.20in.20messages/near/1410903
+  //
+  // With the slash, `href` is  a "path-absolute-URL string". Without the
+  // slash, it's a "path-relative-URL string" meant as relative to the root
+  // path /. By using the URL parser with the root path as a base
+  // (auth.realm should point to the root path), we normalize away that
+  // difference: `.pathname` on a URL object gives the path-absolute string
+  // starting with a slash.
```

Comment:
Hmm and there's one other difference between `href` and `pathname`, even beyond the off-realm cases discussed in the previous comment: `pathname` will drop any query and fragment that might have been in `href`.

That's also a good change -- a query could potentially pass unwanted extra parameters to the API endpoint we're talking to, and a URL fragment isn't meant to go to a server at all.

Reply:
Oh right, thanks for pointing this out!